diff --git a/metatag.admin.inc b/metatag.admin.inc
index 649861d..0051382 100644
--- a/metatag.admin.inc
+++ b/metatag.admin.inc
@@ -776,6 +776,45 @@ function metatag_admin_settings_form() {
     '#default_value' => variable_get('metatag_token_sanitize', FALSE),
   );
 
+  $form['entity_type_groups'] = array(
+    '#type' => 'fieldset',
+    '#title' => t('Entity type / Group Mapping'),
+    '#description' => t('Identify which metatag groups should be available on which entity type / bundle combination. Unselected groups will not appear on the configuration form for that entity type, reducing the size of the form and increasing performance. If no groups are selected for a type, all groups will appear.'),
+    '#collapsible' => TRUE,
+    '#collapsed' => TRUE,
+  );
+
+  $metatag_info = metatag_get_info();
+  $metatag_groups_options = array_map(function($metag_group) {
+    return $metag_group['label'];
+  }, $metatag_info['groups']);
+
+  $index = 0;
+  foreach (entity_get_info() as $entity_type => $entity_info) {
+    // Only deal with entities that are capable of using meta tags.
+    if (metatag_entity_type_is_suitable($entity_type, $entity_info)) {
+      if (!empty($entity_info['bundles'])) {
+        foreach ($entity_info['bundles'] as $bundle_name => $bundle_info) {
+          $form['entity_type_groups'][$index] = array(
+            '#type' => 'fieldset',
+            '#title' => $entity_type . ' - ' . $bundle_name,
+            '#collapsible' => TRUE,
+            '#collapsed' => TRUE,
+          );
+
+          $key = 'metatag_entity_type_groups_' . $entity_type . '_' . $bundle_name;
+          $form['entity_type_groups'][$index][$key] = array(
+            '#type' => 'checkboxes',
+            '#options' => $metatag_groups_options,
+            '#default_value' => variable_get($key, []),
+          );
+
+          $index++;
+        }
+      }
+    }
+  }
+
   // Extra submission logic.
   $form['#submit'][] = 'metatag_admin_settings_form_submit';
 
diff --git a/metatag.install b/metatag.install
index 7e969d2..a98f994 100644
--- a/metatag.install
+++ b/metatag.install
@@ -472,6 +472,7 @@ function metatag_uninstall() {
     if (!empty($entity_info['bundles'])) {
       foreach ($entity_info['bundles'] as $bundle_name => $bundle_info) {
         variable_del('metatag_enable_' . $entity_type . '__' . $bundle_name);
+        variable_del('metatag_entity_type_groups_' . $entity_type . '_' . $bundle_name);
       }
     }
   }
diff --git a/metatag.module b/metatag.module
index a441283..5504a74 100644
--- a/metatag.module
+++ b/metatag.module
@@ -1419,6 +1419,17 @@ function metatag_metatags_form(array &$form, $instance, array $metatags = array(
   // This will be used later.
   $group_metatag_access = array();
 
+  $entity_type_groups = array();
+  // Attempt to extract entity type and bundle to get entity groups setting.
+  // Format: {$entity_type}:{$bundle}.
+  $instance_parts = explode(':', $instance);
+  if (count($instance_parts) > 1) {
+    $entity_type = $instance_parts[0];
+    $bundle_name = $instance_parts[1];
+    $key = 'metatag_entity_type_groups_' . $entity_type . '_' . $bundle_name;
+    $entity_type_groups = array_filter(variable_get($key, array()));
+  }
+
   // Build the form for each metatag.
   foreach ($info['tags'] as $metatag => $metatag_info) {
     // @todo Replace context matching with hook_metatag_access().
@@ -1468,9 +1479,14 @@ function metatag_metatags_form(array &$form, $instance, array $metatags = array(
         '#parents' => array('metatags', $langcode, $metatag),
       );
 
+      // If entity type groups is non-empty we should only show the groups that
+      // is selected.
+      if (!empty($entity_type_groups) && !isset($entity_type_groups[$group_key])) {
+        $form['metatags'][$langcode][$group_key]['#access'] = FALSE;
+      }
       // Hide the fieldset itself if there is not at least one of the meta tag
       // fields visible.
-      if (variable_get('metatag_extended_permissions', FALSE)) {
+      elseif (variable_get('metatag_extended_permissions', FALSE)) {
         $form['metatags'][$langcode][$group_key]['#access'] = count(element_get_visible_children($form['metatags'][$langcode][$group_key])) > 0;
       }
       else {
