diff --git a/includes/date_repeat_entity.update.inc b/includes/date_repeat_entity.update.inc
index 47784a7..6cec3b0 100644
--- a/includes/date_repeat_entity.update.inc
+++ b/includes/date_repeat_entity.update.inc
@@ -202,13 +202,13 @@ function date_repeat_entity_repeating_date_has_changed($original_entity, $update
       $repeating_date_field_name = $repeating_date_field['field_name'];
 
       // Get original date field properties.
-      $original_date = $original_wrapper->{$repeating_date_field_name}[0]->value();
+      $original_date = $original_wrapper->$repeating_date_field_name->value();
       $original_date_start_value = $original_date['value'];
       $original_date_end_value = $original_date['value2'];
       $original_rrule = $original_date['rrule'];
 
       // Get updated date field properties.
-      $updated_date = $updated_wrapper->{$repeating_date_field_name}[0]->value();
+      $updated_date = $updated_wrapper->$repeating_date_field_name->value();
       $updated_date_start_value = $updated_date['value'];
       $updated_date_end_value = $updated_date['value2'];
       $updated_rrule = $updated_date['rrule'];
